File: src/internal/biased-random.php
<?php
require_once 'init.php';

function random(): float {
    return mt_rand() / mt_getrandmax();
}

function getBiasedRandomIndex(array $biases): int {
    /* example
    biases = [ 1.0, 0.5, 2.0, 1.25, 1.0 ]
    expected ranges = [0-1.0)[1.0-1.5)[1.5-3.5)[3.5-4.75)[4.75-5.75)
    if random = 0.0   => [0] 0.0 >= 1.0 False => return 0
    if random = 0.6   => [0] 0.6 >= 1.0 False => return 0
    if random = 4.5   => [0] 4.5 >= 1.0 True [1] 4.5 >= 1.5 True [2] 4.5 >= 3.5 True [3] 4.5 >= 4.75 False => return 3
    if random = 1.5   => [0] 1.5 >= 1.0 True [1] 1.5 >= 1.5 True [2] 1.5 >= 3.5 False => return 2
    if random = 5.74  => [0] 5.74 >= 1.0 [1] 5.74 >= 1.5 [2] >= 3.5 [3] >=4.75 [4] 5.74 >= 5.75 False => return 4
    */

    $rightRangeBordersExclusive = array();
    $border = 0.0;
    foreach ($biases as $bias) {
        $border += $bias;
        $rightRangeBordersExclusive []= $border;
    }

    $random = random() * $border;

    $index = 0;
    while ($random >= $rightRangeBordersExclusive[$index]) {
        $index++;
    }
    return $index;
}

function getMultipleBiasedRandomElements(array $array, array $biases, int $count) {
    if (count($array) <= $count) {
        // nothing to select, just randomize the input array
        $copy = $array;
        shuffle($copy);
        return $copy;
    }
    $result = array();
    $arrayCopy = $array;
    $biasesCopy = $biases;
    for ($i=0; $i<$count; $i++) {
        $randomIndex = getBiasedRandomIndex($biasesCopy);
        $selectedElement = $arrayCopy[$randomIndex];
        $result []= $selectedElement;
        array_splice($arrayCopy, $randomIndex, 1);
        array_splice($biasesCopy, $randomIndex, 1);
    }
    return $result;
}

File: src/main/kotlin/pl/zarajczyk/repetytor/service/BiasedRandom.kt
package pl.zarajczyk.repetytor.service

import kotlin.random.Random

object BiasedRandom {
    
    fun random(): Double = Random.nextDouble()
    
    fun getBiasedRandomIndex(biases: List<Double>): Int {
        val rightRangeBordersExclusive = mutableListOf<Double>()
        var border = 0.0
        for (bias in biases) {
            border += bias
            rightRangeBordersExclusive.add(border)
        }
        
        val random = random() * border
        
        var index = 0
        while (random >= rightRangeBordersExclusive[index]) {
            index++
        }
        return index
    }
    
    fun <T> getMultipleBiasedRandomElements(array: List<T>, biases: List<Double>, count: Int): List<T> {
        if (array.size <= count) {
            return array.shuffled()
        }
        
        val result = mutableListOf<T>()
        val arrayCopy = array.toMutableList()
        val biasesCopy = biases.toMutableList()
        
        for (i in 0 until count) {
            val randomIndex = getBiasedRandomIndex(biasesCopy)
            val selectedElement = arrayCopy[randomIndex]
            result.add(selectedElement)
            arrayCopy.removeAt(randomIndex)
            biasesCopy.removeAt(randomIndex)
        }
        
        return result
    }
}
